File: ServerCode/business-application-service/src/main/js/utils/web3Utils.js
const { ethers } = require("ethers");
const createBaseRequest = require("../common/http-common");

const CCTOKEN_ABI = require("../contracts/abi/CCToken.json");
const FACTORY_ABI = require("../contracts/abi/CCProjectFactory.json");
const PROJECT_ABI = require("../contracts/abi/CCProject.json");

let CCTOKEN_ADDRESS = null;
let FACTORY_ADDRESS = null;
let contractAddressesPromise = null;

/**
 * Fetch contract addresses from backend API
 * These are read from application.properties server-side
 */
const fetchContractAddresses = async () => {
  if (contractAddressesPromise) {
    return contractAddressesPromise;
  }

  contractAddressesPromise = (async () => {
    try {
      const response = await createBaseRequest().get("/api/blockchain/config/contracts");
      CCTOKEN_ADDRESS = response.data.tokenAddress;
      FACTORY_ADDRESS = response.data.factoryAddress;
      console.log("Contract addresses loaded from backend:", {
        token: CCTOKEN_ADDRESS,
        factory: FACTORY_ADDRESS
      });
      return { token: CCTOKEN_ADDRESS, factory: FACTORY_ADDRESS };
    } catch (error) {
      console.error("Error fetching contract addresses from backend:", error);
      // Fallback to hardcoded values if API call fails
      CCTOKEN_ADDRESS = "0x5FbDB2315678afecb367f032d93F642f64180aa3";
      FACTORY_ADDRESS = "0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512";
      console.warn("Using fallback contract addresses");
      return { token: CCTOKEN_ADDRESS, factory: FACTORY_ADDRESS };
    }
  })();

  return contractAddressesPromise;
};

/**
 * Fetching contract addresses from backend if not already loaded
 */
const getContractAddresses = async () => {
  if (!CCTOKEN_ADDRESS || !FACTORY_ADDRESS) {
    await fetchContractAddresses();
  }
  return {
    token: CCTOKEN_ADDRESS,
    factory: FACTORY_ADDRESS
  };
};

const getCCTokenBalance = async () => {
  try {
    // To ensure contract addresses are loaded first
    const addresses = await getContractAddresses();
    const tokenAddress = addresses.token;
    
    await window.ethereum.request({ method: "eth_requestAccounts" });
    
    // Get chain ID directly from MetaMask first
    const chainIdHex = await window.ethereum.request({ method: "eth_chainId" });
    const chainId = parseInt(chainIdHex, 16);
    const expectedChainId = 31337; // Hardhat local network
    
    console.log("MetaMask Chain ID:", chainId, "Expected:", expectedChainId);
    
    if (chainId !== expectedChainId) {
      // Try to switch network automatically
      try {
        await window.ethereum.request({
          method: 'wallet_switchEthereumChain',
          params: [{ chainId: '0x7A69' }], // 31337 in hex
        });
        // Wait a moment for network to switch
        await new Promise(resolve => setTimeout(resolve, 1000));
        // Check again
        const newChainIdHex = await window.ethereum.request({ method: "eth_chainId" });
        const newChainId = parseInt(newChainIdHex, 16);
        if (newChainId !== expectedChainId) {
          throw new Error(`Please switch to Hardhat Local network (Chain ID: ${expectedChainId}). Current network: Chain ID ${newChainId}`);
        }
      } catch (switchError) {
        if (switchError.code === 4902) {
          try {
            await window.ethereum.request({
              method: 'wallet_addEthereumChain',
              params: [{
                chainId: '0x7A69', // 31337 in hex
                chainName: 'Hardhat Local',
                nativeCurrency: {
                  name: 'Ethereum',
                  symbol: 'ETH',
                  decimals: 18
                },
                rpcUrls: ['http://127.0.0.1:8545'],
                blockExplorerUrls: null
              }],
            });
            // Wait a moment for network to be added and switched
            await new Promise(resolve => setTimeout(resolve, 1000));
          } catch (addError) {
            throw new Error(`Please manually switch to Hardhat Local network (Chain ID: ${expectedChainId}). Current network: Chain ID ${chainId}. Error: ${addError.message}`);
          }
        } else {
          throw new Error(`Please switch to Hardhat Local network (Chain ID: ${expectedChainId}). Current network: Chain ID ${chainId}. Error: ${switchError.message}`);
        }
      }
    }
    
    const provider = await new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    provider.send("eth_requestAccounts", []);
    const signer = provider.getSigner();
    
    // Check if contract exists at address
    const code = await provider.getCode(tokenAddress);
    if (code === "0x" || code === "0x0") {
      throw new Error(`Token contract not found at address ${tokenAddress}. Make sure contracts are deployed to Hardhat network.`);
    }
    
    const ccTokenContract = new ethers.Contract(
      tokenAddress,
      CCTOKEN_ABI,
      signer
    );
    const walletAddress = await signer.getAddress();
    const decimals = await ccTokenContract.decimals();
    const tokens = await ccTokenContract.balanceOf(walletAddress);
    return await ethers.utils.formatEther(tokens, decimals);
  } catch (ex) {
    console.error("Error getting token balance:", ex);
    const errorMessage = ex.message || "Failed to get token balance. Make sure Hardhat node is running and MetaMask is connected to Localhost 8545 (Chain ID: 31337).";
    throw errorMessage;
  }
};

const appendZeros = (amount, count) => {
  amount = amount.toString();
  for (let i = 0; i < count; i++) {
    amount = amount + "0";
  }
  return amount;
};

const getWalletAddress = async () => {
  try {
    await window.ethereum.request({ method: "eth_requestAccounts" });
    const provider = await new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    provider.send("eth_requestAccounts", []);
    const signer = provider.getSigner();
    return await signer.getAddress();
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const approveFunding = async (projectAddress, amount) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const ccTokenContract = new ethers.Contract(
      addresses.token,
      CCTOKEN_ABI,
      signer
    );
    const tx = await ccTokenContract.increaseAllowance(
      projectAddress,
      appendZeros(amount, 18)
    );
    await tx.wait();
    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const createCCProjectContract = async (formData) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const ownerAddress = formData.get("walletAddress");
    const listing = JSON.parse(formData.get("listing"));
    const vinNumber = listing.vin;
    const make = listing.make;
    const model = listing.model;
    const ccpg = appendZeros(listing.ccpg, 18);
    const fundingGoal = appendZeros(listing.fundingGoal, 18);

    let projectAddresses = await factoryContract.getProjects();
    const projectCount = projectAddresses.length;

    const tx = await factoryContract.createNewProject(
      vinNumber,
      make,
      model,
      ccpg,
      fundingGoal,
      ownerAddress
    );
    await tx.wait();
    console.log(tx.hash);

    projectAddresses = await factoryContract.getProjects();
    while (projectAddresses.length == projectCount) {
      projectAddresses = await factoryContract.getProjects();
    }

    projectAddress = projectAddresses[projectAddresses.length - 1];
    console.log(projectAddress);

    return projectAddress;
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const approveCCProjectListing = async (projectAddress) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
  
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const tx = await factoryContract.approveProject(projectAddress);
    await tx.wait();
    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const saveEstCCProjectListing = async (
  projectAddress,
  valueEstimate,
  repairEstimate
) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const tx = await factoryContract.saveEstimations(
      projectAddress,
      valueEstimate,
      repairEstimate
    );
    await tx.wait();

    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const setRestorerCCProjectListing = async (
  projectAddress,
  restorerAddress,
  fundingGoal
) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const tx = await factoryContract.setRestorer(
      projectAddress,
      restorerAddress,
      appendZeros(fundingGoal, 18)
    );
    await tx.wait();

    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const assignRestorationCCProjectListing = async (projectAddress) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const tx = await factoryContract.assignForRestoration(projectAddress);
    await tx.wait();

    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const openAuctionCCProjectListing = async (projectAddress) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const tx = await factoryContract.setAuctionOpen(projectAddress);
    await tx.wait();

    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const setBuyerCCProjectListing = async (projectAddress, buyerAddress) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const tx = await factoryContract.setBuyer(projectAddress, buyerAddress);
    await tx.wait();

    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const redistributeCCProjectListing = async (projectAddress) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const tx = await factoryContract.redistribute(projectAddress);
    await tx.wait();

    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const updateCCProjectListing = async (projectAddress, formData) => {
  try {
    // To ensure contract addresses are loaded
    const addresses = await getContractAddresses();
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const factoryContract = new ethers.Contract(
      addresses.factory,
      FACTORY_ABI,
      signer
    );

    const listing = JSON.parse(formData.get("listing"));
    const vinNumber = listing.vin;
    const make = listing.make;
    const model = listing.model;
    const ccpg = appendZeros(listing.ccpg, 18);
    const fundingGoal = appendZeros(listing.fundingGoal, 18);

    const tx = await factoryContract.editProjectDetails(
      projectAddress,
      vinNumber,
      make,
      model,
      ccpg,
      fundingGoal
    );
    await tx.wait();
    console.log(tx);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const investInCCProjectListing = async (projectAddress, amount) => {
  try {
    await approveFunding(projectAddress, amount);
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const projectContract = new ethers.Contract(
      projectAddress,
      PROJECT_ABI,
      signer
    );

    const tx = await projectContract.acceptFunds(appendZeros(amount, 18));
    const receipt = await tx.wait();

    console.log("Investment transaction hash:", receipt.transactionHash);
    // Return transaction hash for backend synchronization
    return receipt.transactionHash;
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

const bidForCCProjectListing = async (projectAddress, amount) => {
  try {
    await approveFunding(projectAddress, amount);
    
    const provider = new ethers.providers.Web3Provider(
      window.ethereum,
      "any"
    );
    const signer = provider.getSigner();
    const projectContract = new ethers.Contract(
      projectAddress,
      PROJECT_ABI,
      signer
    );

    const tx = await projectContract.addNewAuctionBid(appendZeros(amount, 18));
    await tx.wait();

    console.log(tx.hash);
  } catch (ex) {
    console.log(ex);
    throw "Web3 Error!";
  }
};

module.exports = {
  getCCTokenBalance,
  getWalletAddress,
  approveFunding,
  createCCProjectContract,
  updateCCProjectListing,
  approveCCProjectListing,
  saveEstCCProjectListing,
  setRestorerCCProjectListing,
  assignRestorationCCProjectListing,
  openAuctionCCProjectListing,
  setBuyerCCProjectListing,
  redistributeCCProjectListing,
  investInCCProjectListing,
  bidForCCProjectListing,
};

File: ServerCode/business-application-service/src/main/java/com/company/utils/Web3Utils.java
package com.company.utils;

import org.web3j.abi.FunctionEncoder;
import org.web3j.abi.TypeReference;
import org.web3j.abi.datatypes.Address;
import org.web3j.abi.datatypes.Function;
import org.web3j.abi.datatypes.Type;
import org.web3j.abi.datatypes.Uint;
import org.web3j.abi.datatypes.generated.Uint256;
import org.web3j.contracts.eip20.generated.ERC20;
import org.web3j.crypto.Credentials;
import org.web3j.crypto.RawTransaction;
import org.web3j.crypto.TransactionEncoder;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.core.DefaultBlockParameterName;
import org.web3j.protocol.core.methods.request.Transaction;
import org.web3j.protocol.core.methods.response.EthGetBalance;
import org.web3j.protocol.core.methods.response.EthSendTransaction;
import org.web3j.protocol.http.HttpService;
import org.web3j.tx.RawTransactionManager;
import org.web3j.tx.TransactionManager;
import org.web3j.utils.Convert;
import org.web3j.utils.Numeric;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Web3 utility class for blockchain interactions
 * Handles contract interactions with CCToken, CCProjectFactory, and CCProject contracts
 */
public class Web3Utils {
    
    private static final String CCTOKEN_ADDRESS = "0x60c5eb023F7778031F542ece23dff7165a515CC0";
    private static final String FACTORY_ADDRESS = "0xc2B45f7DfCf4Fb869133cC35BEdaA9c63D95Fda3";
    private static final String RPC_URL = "http://localhost:8545"; // Update with your RPC endpoint
    
    private Web3j web3j;
    private Credentials credentials;
    
    public Web3Utils(String rpcUrl, String privateKey) {
        this.web3j = Web3j.build(new HttpService(rpcUrl));
        this.credentials = Credentials.create(privateKey);
    }
    
    public Web3Utils() {
        this.web3j = Web3j.build(new HttpService(RPC_URL));
    }

    /**
     * Get CC Token balance for a given address
     */
    public String getCCTokenBalance(String walletAddress) throws Exception {
        try {
            EthGetBalance balance = web3j.ethGetBalance(walletAddress, DefaultBlockParameterName.LATEST).send();
            BigDecimal tokenBalance = Convert.fromWei(new BigDecimal(balance.getBalance()), Convert.Unit.ETHER);
            return tokenBalance.toString();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Append zeros to amount (for decimal conversion)
     */
    public static BigInteger appendZeros(String amount, int count) {
        String result = amount;
        for (int i = 0; i < count; i++) {
            result = result + "0";
        }
        return new BigInteger(result);
    }

    /**
     * Append zeros to amount (for BigInteger)
     */
    public static BigInteger appendZeros(BigInteger amount, int count) {
        return appendZeros(amount.toString(), count);
    }

    /**
     * Approve funding for a project
     */
    public String approveFunding(String projectAddress, String amount) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            // Create approve function
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            inputParameters.add(new Uint256(appendZeros(amount, 18)));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            outputParameters.add(new TypeReference<>() {});
            
            Function function = new Function("increaseAllowance", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Create a new CC Project contract
     */
    public String createCCProjectContract(String walletAddress, String vin, String make, String model, 
                                         String ccpg, String fundingGoal) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            // Create createNewProject function
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new org.web3j.abi.datatypes.Utf8String(vin));
            inputParameters.add(new org.web3j.abi.datatypes.Utf8String(make));
            inputParameters.add(new org.web3j.abi.datatypes.Utf8String(model));
            inputParameters.add(new Uint256(appendZeros(ccpg, 18)));
            inputParameters.add(new Uint256(appendZeros(fundingGoal, 18)));
            inputParameters.add(new Address(walletAddress));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("createNewProject", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Approve a CC Project listing
     */
    public String approveCCProjectListing(String projectAddress) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("approveProject", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Save estimations for a project
     */
    public String saveEstCCProjectListing(String projectAddress, BigInteger valueEstimate, 
                                         BigInteger repairEstimate) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            inputParameters.add(new Uint256(valueEstimate));
            inputParameters.add(new Uint256(repairEstimate));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("saveEstimations", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Set restorer for a project
     */
    public String setRestorerCCProjectListing(String projectAddress, String restorerAddress, 
                                             String fundingGoal) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            inputParameters.add(new Address(restorerAddress));
            inputParameters.add(new Uint256(appendZeros(fundingGoal, 18)));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("setRestorer", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Assign project for restoration
     */
    public String assignRestorationCCProjectListing(String projectAddress) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("assignForRestoration", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Open auction for a project
     */
    public String openAuctionCCProjectListing(String projectAddress) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("setAuctionOpen", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Set buyer for a project
     */
    public String setBuyerCCProjectListing(String projectAddress, String buyerAddress) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            inputParameters.add(new Address(buyerAddress));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("setBuyer", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Redistribute funds for a project
     */
    public String redistributeCCProjectListing(String projectAddress) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("redistribute", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Update project listing details
     */
    public String updateCCProjectListing(String projectAddress, String vin, String make, String model,
                                        String ccpg, String fundingGoal) throws Exception {
        try {
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Address(projectAddress));
            inputParameters.add(new org.web3j.abi.datatypes.Utf8String(vin));
            inputParameters.add(new org.web3j.abi.datatypes.Utf8String(make));
            inputParameters.add(new org.web3j.abi.datatypes.Utf8String(model));
            inputParameters.add(new Uint256(appendZeros(ccpg, 18)));
            inputParameters.add(new Uint256(appendZeros(fundingGoal, 18)));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("editProjectDetails", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Invest in a project
     */
    public String investInCCProjectListing(String projectAddress, String amount) throws Exception {
        try {
            // First approve funding
            approveFunding(projectAddress, amount);
            
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Uint256(appendZeros(amount, 18)));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("acceptFunds", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Place a bid for a project in auction
     */
    public String bidForCCProjectListing(String projectAddress, String amount) throws Exception {
        try {
            // First approve funding
            approveFunding(projectAddress, amount);
            
            TransactionManager transactionManager = new RawTransactionManager(web3j, credentials);
            
            List<Type> inputParameters = new ArrayList<>();
            inputParameters.add(new Uint256(appendZeros(amount, 18)));
            
            List<TypeReference<?>> outputParameters = new ArrayList<>();
            
            Function function = new Function("addNewAuctionBid", inputParameters, outputParameters);
            String encodedFunction = FunctionEncoder.encode(function);
            
            EthSendTransaction response = web3j.ethSendRawTransaction(encodedFunction).send();
            
            if (response.hasError()) {
                throw new RuntimeException("Transaction failed: " + response.getError().getMessage());
            }
            
            return response.getTransactionHash();
        } catch (Exception ex) {
            throw new RuntimeException("Web3 Error: " + ex.getMessage(), ex);
        }
    }

    /**
     * Get contract addresses
     */
    public String getCCTokenAddress() {
        return CCTOKEN_ADDRESS;
    }

    public String getFactoryAddress() {
        return FACTORY_ADDRESS;
    }
}
